### ISSUE TITLE
Help with hyper-parametric tuning for yolo11

### ISSUE THREAD

**Issue by sarabhan** [ultralytics/ultralytics#22720]
Hi, i am training a yolo11s model on my custon dataset for 5 classes. however, the dataset is unbalanced. 

this is the distribution of my dataset: 
`{0: 2116, 3: 2362, 2: 741, 4: 1073, 1: 94}`

and this is my data.yaml file
```
train: ../train/images
val: ../valid/images
test: ../test/images

nc: 5
names: ['helmet', 'no-helmet', 'no-vest', 'person', 'vest']

```
i know image augmentation and adding more images to the underrepresented classes is the best solution, but i do not have much time. 

is there any way i can tune the hyper-parameters of yolo11 before training? or assign class weights to help underrepresented classes? 

if no, is there any workaround for this?
help!!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @sarabhan, thank you for your interest in Ultralytics and YOLO11 🚀! This is an automated response to help get things started. An Ultralytics engineer will review your issue and assist you further soon 😊

Since your question is about custom training and hyperparameter tuning on an imbalanced dataset, please share a bit more context so we can better understand your setup when we follow up:
- Your training command (Python or CLI) and any custom hyperparameters you are already using  
- A short snippet of your training logs (including metrics like mAP, precision, recall)  
- A few example images and labels from your underrepresented classes  

In the meantime, we recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report (for example, if something crashes or behaves unexpectedly during training), please also provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (which it appears to be), please provide as much information as possible, including:
- Your full `data.yaml` (you already shared the main parts 👍)  
- Training logs and final metrics  
- Any custom settings you are using in your training command  
- Confirmation that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)

Join the Ultralytics community where it suits you best:
- For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧  
- Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/)  
- Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community  

## Upgrade

Before we dive deeper, please ensure you are on the latest version of the `ultralytics` package, with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) installed in a **Python>=3.8** environment with **PyTorch>=1.8** to make sure you benefit from the latest YOLO11 features and fixes:

```bash
pip install -U ultralytics
```

## Environments

YOLO (including YOLO11) may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Thanks for sharing the class counts; that imbalance is significant, especially for `no-helmet` with only 94 samples.  

There isn’t a user-facing `class_weights` argument in YOLO11 right now, so the simplest effective workaround is to rebalance at the data level: duplicate or symlink the `no-helmet` images (with their labels) into `train/` a few times so that class 1 appears more often during training, while keeping `val`/`test` unchanged.  

If you have a bit of extra compute, you can also try a small hyperparameter search to adjust things like `cls` and augmentations for your dataset using `model.tune()` as described in the [Hyperparameter Tuning guide](https://docs.ultralytics.com/guides/hyperparameter-tuning/):  

```python
from ultralytics import YOLO

model = YOLO("yolo11s.pt")

search_space = {
    "cls": (0.5, 4.0),      # classification loss weight
    "mosaic": (0.0, 1.0),
    "mixup": (0.0, 1.0),
}

model.tune(
    data="data.yaml",
    epochs=30,
    iterations=100,  # keep this small if you’re short on time
    space=search_space,
    plots=False,
    save=False,
    val=False,
)
```  

If you can share how the minority class performs after oversampling (precision/recall for `no-helmet`), we can suggest more targeted tweaks.

**Comment by sarabhan** [comment]
considering the automatic searching approach for the optimal hyper parameter values, i read the guide and i am aware that the search will result in a .yaml file that has all the optimal hyper-parameter values. however, how do i use those values for my training? 

is there a specific argument i must pass referencing the hyperparameter.yaml file in the train command?

also, for the automatic parameter search, do i need to pass in any other arguments related to the train & test directories?

**Comment by glenn-jocher** [comment]
Great questions.  

1) After tuning, you can plug `best_hyperparameters.yaml` directly into training via the `cfg` argument; for example:  

```python
from ultralytics import YOLO

model = YOLO("yolo11s.pt")

model.train(
    data="data.yaml",  # your dataset YAML
    cfg="runs/detect/tune/best_hyperparameters.yaml",  # path from the tuning run
    epochs=100,
)
```  

The hyperparameters in that file will be merged into your normal training args.  

2) For `model.tune(...)` you only need to pass `data="data.yaml"`; the `train/val/test` folders are already defined inside your dataset YAML, so no extra directory arguments are required unless you want to override something.  

You can find the relevant arguments for both `train()` and `tune()` in the [model API reference](https://docs.ultralytics.com/reference/engine/model/).

**Comment by sarabhan** [comment]
thank you so much!

**Comment by glenn-jocher** [comment]
Glad it helped; if you run into anything else while using `model.tune()` or during the main `train()` run, just update this issue with your command and logs and we’ll take a look.
